# Servo Config for Sheldon
# Limits defined here are absolute limits (before something breaks).  Limits in code/Moveit should be smaller than these.

## MODIFIED FOR CORRECT SPEED
## SERVO SPEEDS USED:
# MX28:   55 RPM --> 5.8 rad/sec
# MX64:   63 RPM --> 6.6 rad/sec
# MX-106: 45 RPM --> 4.6 rad/sec
# MX-106 (gear reduced for shoulder): 45 RPM / 2 --> 2.3 rad/sec


## WARNING!  IF YOU ADD SERVOS, CHANGE "max_motor_id" in "servos.launch"

dynamixels: [

  'head_sidetilt_joint', 'head_tilt_joint', 'head_pan_joint',
  # 'chest_camera_joint',

  'right_arm_shoulder_rotate_joint',  'right_arm_shoulder_lift_joint', 
  'right_arm_elbow_rotate_joint',     'right_arm_elbow_bend_joint',
  'right_arm_wrist_bend_joint',       'right_arm_wrist_rotate_joint',
  'right_arm_gripper_finger_joint', 

  'left_arm_shoulder_rotate_joint',   'left_arm_shoulder_lift_joint',
  'left_arm_elbow_rotate_joint',      'left_arm_elbow_bend_joint', 
  'left_arm_wrist_bend_joint',        'left_arm_wrist_rotate_joint',
  'left_arm_gripper_finger_joint']


controllers: [

  'head_sidetilt_joint', 'head_tilt_joint', 'head_pan_joint',
  #'chest_camera_joint',

  'right_arm_shoulder_rotate_joint',  'right_arm_shoulder_lift_joint', 
  'right_arm_elbow_rotate_joint',     'right_arm_elbow_bend_joint',
  'right_arm_wrist_bend_joint',       'right_arm_wrist_rotate_joint',
  'right_arm_gripper_finger_joint', 

  'left_arm_shoulder_rotate_joint',   'left_arm_shoulder_lift_joint',
  'left_arm_elbow_rotate_joint',      'left_arm_elbow_bend_joint', 
  'left_arm_wrist_bend_joint',        'left_arm_wrist_rotate_joint',
  'left_arm_gripper_finger_joint']


# Head
head_sidetilt_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_sidetilt_joint
    joint_speed: 5.8
    motor:
        id: 2
        init: 2047 
        min: 1600     # sidetilt left
        max: 2504  # sidetilt right

head_tilt_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_tilt_joint
    joint_speed: 5.8
    motor:
        id: 3
        init: 2047
        min: 1541  # look up
        max: 2792  # look down

head_pan_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_pan_joint
    joint_speed: 6.6
    motor:
        id: 4
        init: 2047
        min: 921    # look right
        max: 3195   # look left

# TODO chest_camera_joint:
#    motor:
#        id: 5


# Right Arm
right_arm_shoulder_lift_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_shoulder_lift_joint
    joint_speed: 4.6
    motor:
        id: 6
        init: 2065 # changed 2074
        min: 1999
        max: 3445

right_arm_elbow_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_elbow_rotate_joint
    joint_speed: 6.6
    motor:
        id: 7
        init: 2100  # 2047
        min: 300
        max: 4095

right_arm_elbow_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_elbow_bend_joint
    joint_speed: 6.6
    motor:
        id: 8
        init: 2047
        min: 1189  
        max: 4095  

right_arm_wrist_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_wrist_bend_joint
    joint_speed: 6.6
    motor:
        id: 9
        init: 2047
        min: 1189  
        max: 4095  

right_arm_wrist_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_wrist_rotate_joint
    joint_speed: 5.8
    motor:
        id: 10
        init: 2047
        min: 64
        max: 3270

right_arm_gripper_finger_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_gripper_finger_joint
    joint_speed: 5.8
    motor:
        id: 11
        init: 2047
        min: 2184
        max: 0    # reversed


#Left Arm
left_arm_shoulder_lift_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_shoulder_lift_joint
    joint_speed: 4.6
    motor:
        id: 12
        init: 2200 # changed
        min: 2140
        max: 3586

left_arm_elbow_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_elbow_rotate_joint
    joint_speed: 6.6
    motor:
        id: 13
        init: 2047
        min: 4095 # reversed
        max: 300  # turn outward = positive

left_arm_elbow_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_elbow_bend_joint
    joint_speed: 6.6
    motor:
        id: 14
        init: 2047
        min: 1189
        max: 4095

left_arm_wrist_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_wrist_bend_joint
    joint_speed: 6.6
    motor:
        id: 15
        init: 2047
        min: 4095    # reversed
        max: 1189

left_arm_wrist_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_wrist_rotate_joint
    joint_speed: 5.8
    motor:
        id: 16
        init: 2047
        min: 4095  # reversed
        max: 752   # turn outward = positive

left_arm_gripper_finger_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_gripper_finger_joint
    joint_speed: 5.8
    motor:
        id: 17
        init: 2047
        min: 1894
        max: 4095  


# Shoulder Motors
right_arm_shoulder_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_arm_shoulder_rotate_joint
    joint_speed: 2.3
    motor:
        id: 18
        init: 2047  #2047 default
        min: 4095  # reversed
        max: 0     # positive = forward

left_arm_shoulder_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_arm_shoulder_rotate_joint
    joint_speed: 2.3
    motor:
        id: 19
        init: 2047
        min: 0
        max: 4095


